src/entity/find-entities.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import {Document} from '../parser/document';
+import {Descriptor} from '../ast/ast';
+import {EntityFinder} from './entity-finder';
+
+export async function findEntities(
+    document: Document<any, any>,
+    finders: EntityFinder<any, any, any>[]): Promise<Descriptor[]> {
+
+  // Finders register a visitor to run via the `visit` callback passed to
+  // `findEntities()`. We run these visitors in a batch, then pass control back
+  // to the `findEntities` methods by resolving a single Promise return for
+  // all calls to visit() in a batch when the visitors have run.
+  // Then we repeat if any visitors have registered new visitors.
+
+  let entities: Descriptor[] = [];
+
+  // Resolves Promises returned by visit() calls
+  let batchDone: () => void;
+
+  // Promise returned by visit()
+  let visitorsPromise: Promise<void>;
+
+  // Current batch of visitors
+  let visitors: any[];
+
+  // A Promise that runs the next batch of visitors in a microtask
+  let runner: Promise<void>;
+
+  let visitError: any;
+
+  // Initializes the current batch running state
+  function setup() {
+    visitorsPromise = new Promise<void>((resolve, reject) => {
+      batchDone = resolve;
+    });
+    visitors = [];
+    runner = null;
+  }
+
+  // Runs the current batch of visitors
+  function runVisitors() {
+    // Record the current state so that any new visitors are enqueued into
+    // a fresh batch.
+    let currentVisitors = visitors;
+    let currentDoneCallback = batchDone;
+    let currentVisitorsPromise = visitorsPromise;
+    setup();
+
+    try {
+      document.visit(currentVisitors);
+    } catch (error) {
+      visitError = error;
+    }
+
+    // Let `findEntities` continue after calls to visit().then()
+    currentDoneCallback();
+  };
+
+  // The callback passed to `findEntities()`
+  function visit(visitor: any) {
+    visitors.push(visitor);
+    if (!runner) {
+      runner = Promise.resolve().then(runVisitors);
```

Comment:
I'm pretty sympathetic to this point, and considered it. The reason I went down this path is because I felt the complexity could be localized here, and documented and tested well, while keeping `findEntities()` implementations simpler. Once this is written and burned in, it will need much less attention than the many upcoming `findEntities()` implementations, possibly written by many different authors, and dependent on different libraries. Especially because `findEntities()` will be part of the eventual user-exposed plugin API, it should have a simple contract. This approach preserves `visit()` as a pretty straightforward async call, even if it's complex behind the scenes.

There are other options that also preserve simplicity in `findEntities()`: `visit()` could be synchronous, `visit()` could be async but not batch, or `findEntities()` could be two phase: it returns a set of visitors, and then has some kind of finish callback.

I still think for the moment that the behavior that this implementation gives `findEntities()` is close to ideal, but I could be convinced otherwise.